Calling fmi2SetXXX in event mode

nik5346 opened the pull request:
### Related Issues

#10915

### Purpose

Use interger, boolean and string inputs for fmu simulation.

### Approach

calling fmi2SetXXX functions in event mode. This call sequence does not violate the fmi2 standard.

AnHeuermann commented:
Just to make extra sure this approach in general should work and that the Dymola FMU works I recreated all of it with Dymola and indeed there it is working.

![grafik](https://github.com/OpenModelica/OpenModelica/assets/38031952/d72d12bb-44ca-4740-ae6d-fb171791a01c)

nik5346 commented:
This is expected. This is what you get, when OpenModelica implements the call sequence according to the fmi2 standard and other tools do not.

As I already stated multiple times:
In the FMI standard documentation (version 2.0.4, page 90) the functions fmi2SetReal, fmi2SetInteger, fmi2SetBoolean and fmi2SetString are allowed to be called in event mode. In continuous time mode, however, only the call to fm2SetReal is allowed. 
All other tools ignore this fact. They call fmi2SetReal, fmi2SetInteger, fmi2SetBoolean and fmi2SetString in continuous time mode. 

Your feedthrough fmu generated by Dymola does not trigger the event mode. When simulating your FMU in Dymola, it works, because Dymola calls fmi2SetReal, fmi2SetInteger, fmi2SetBoolean and fmi2SetString in continuous time mode. 

When simulating this FMU in OpenModelica with #10916 it should also WORK, because I fixed the issue by calling  fmi2SetReal, fmi2SetInteger, fmi2SetBoolean and fmi2SetString in continuous time mode.

When simulating this FMU in OpenModelica with #10945 it should NOT WORK, because I fixed the issue by calling  fmi2SetReal, fmi2SetInteger, fmi2SetBoolean and fmi2SetString in event mode only. However your FMU does not trigger the event mode.

The issue is fixed once in #10916 and again in #10945. No changes are needed. Just choose between tool compliant or standard compliant.

Test this issue with the FMU provided: https://de.mathworks.com/help/simulink/ug/fmu-block-calling-sequences.html
This FMU definitly triggers the event mode as you can see in the Code!

AnHeuermann commented:
> Your observations are the exact reason, I originally recommended https://github.com/OpenModelica/OpenModelica/pull/10916 to fix the issue.

In that case I would recommend re-opening https://github.com/OpenModelica/OpenModelica/pull/10916 and we can merge it.

---------------

> Furthermore, don't mix issues from fmu exports from other tools with the call sequence bug in OpenModelica! Test this issue with the simple FMU provided here: https://github.com/nik5346/FeedThroughFmu

> fmi2eventInfo->newDiscreteStatesNeeded = fmi2False;
fmi2eventInfo->nextEventTime = simTime + 0.1;
fmi2eventInfo->nextEventTimeDefined = fmi2True;

Sorry if I'm nitpicking, but that is simply not a good implementation of the FMI standard. The FeedThrough FMU should not trigger events. The next event should not trigger in 0.1 seconds. The FMU doesn't know when the next event will trigger.
When in doubt that if it's a FMU export issue I would recommend the official reference FMUs from the FMI specification. In this case it would be [Feedthrough](https://github.com/modelica/Reference-FMUs/blob/main/Feedthrough/model.c).

If you need a special FMU to get this PR to work I don't see the point of it.

nik5346 commented:
> Sorry if I'm nitpicking, but that is simply not a good implementation of the FMI standard. The FeedThrough FMU should not trigger events. The next event should not trigger in 0.1 seconds. The FMU doesn't know when the next event will trigger.
When in doubt that if it's a FMU export issue I would recommend the official reference FMUs from the FMI specification. In this case it would be [Feedthrough](https://github.com/modelica/Reference-FMUs/blob/main/Feedthrough/model.c).

Sure this is not good design, but strictly speaking it is the only standard compliant way to call fmi2SetInteger. In my opinion this is a fmi2 design flaw. I opened an issue for this flaw: https://github.com/modelica/fmi-standard/issues/1890

> If you need a special FMU to get this PR to work I don't see the point of it.

The point of this PR is not to make the Feedthrough FMU work :). The purpose of this PR is to make a FMU, I am working on, function. This FMU needs to set inputs of all types continuously. Therefor I don't care if fmi2SetInteger is called in continuous time or in event mode only, since I can trigger the event mode from within the FMU, whenever needed.
